src/translate/pddl_parser/parsing_functions.py

```diff
@@ -313,186 +547,168 @@ def parse_task(domain_pddl, task_pddl):
-                    print("Warning: %s is specified twice" % assignment,
-                          "in initial state specification")
-                else:
-                    raise SystemExit("Error in initial state specification\n" +
-                                     "Reason: conflicting assignment for " +
-                                     "%s." % assignment.fluent)
-            else:
-                initial_assignments[assignment.fluent] = assignment
-                initial.append(assignment)
-        elif fact[0] == "not":
-            atom = pddl.Atom(fact[1][0], fact[1][1:])
-            check_atom_consistency(atom, initial_false, initial_true, False)
-            initial_false.add(atom)
+    with context.layer("Parsing task"):
+        define_tag = next(iterator)
+        if define_tag != "define":
+            context.error("Task definition expected to start with '(define ")
+
+        with context.layer("Parsing problem name"):
+            problem_line = next(iterator)
+            context.check_expected_block(
+                problem_line, ["problem"], min_size=2, max_size=2)
+            yield problem_line[1]
+
+        with context.layer("Parsing domain name"):
+            domain_line = next(iterator)
+            context.check_expected_block(
+                domain_line, [":domain"], min_size=2, max_size=2)
+            yield domain_line[1]
+
+        requirements_opt = next(iterator)
+        context.check_expected_block(requirements_opt, [":requirements", ":objects", ":init"])
+        if requirements_opt[0] == ":requirements":
+            requirements = requirements_opt[1:]
+            objects_opt = next(iterator)
         else:
-            atom = pddl.Atom(fact[0], fact[1:])
-            check_atom_consistency(atom, initial_true, initial_false)
-            initial_true.add(atom)
-    initial.extend(initial_true)
-    yield initial
-
-    goal = next(iterator)
-    assert goal[0] == ":goal" and len(goal) == 2
-    yield parse_condition(goal[1], type_dict, predicate_dict)
-
-    use_metric = False
-    for entry in iterator:
-        if entry[0] == ":metric":
-            if entry[1] == "minimize" and entry[2][0] == "total-cost":
-                use_metric = True
-            else:
-                assert False, "Unknown metric."
-    yield use_metric
+            requirements = []
+            objects_opt = requirements_opt
+        yield parse_requirements(context, requirements)
+
+        context.check_expected_block(objects_opt, [":objects", ":init"])
+        if objects_opt[0] == ":objects":
+            with context.layer("Parsing objects"):
+                yield parse_typed_list(context, objects_opt[1:])
+            init = next(iterator)
+        else:
+            yield []
+            init = objects_opt
+
+        yield parse_init(context, init)
+
+        goal = next(iterator)
+        context.check_expected_block(goal, [":goal"])
+        with context.layer("Parsing goal"):
```

Comment:
The code has currently the pattern:
0. We are in context layer L.
1. get the next element from the iterator (still in layer L)
2. Check if this element matches an expected block (in earlier examples, there are more options, e.g. :action or :derived, thus, it makes no sense to construct a layer 'Parsing action or axiom' if the error message is already expecting action or axiom)
3. Once we know the things to parse next, create the context layer and do the parsing.

Reply:
This sounds more like a problem with the other places, where we could use a layer `Checking what next block is` (because it is an open question there). If we would extract a function to parse the goal, I'd include the previous two lines in that function. With this argument, I would also include them in the layer.